CRM/MembershipExtras/Job/OfflineAutoRenewal.php

```diff
@@ -240,49 +234,192 @@ private function getContributionStatusesNameMap() {
+      return  $lineItem['unit_price'];
     }
 
-    return $amount;
+    $optoutUsingLastPriceFieldID = civicrm_api3('CustomField', 'getvalue', [
+      'return' => 'id',
+      'custom_group_id' => 'offline_autorenew_option',
+      'name' => 'optout_last_price_offline_autorenew',
+    ]);
+
+    $lineItemMembership = civicrm_api3('Membership', 'get', [
+      'sequential' => 1,
+      'return' => ["custom_$optoutUsingLastPriceFieldID", 'membership_type_id.minimum_fee'],
+      'id' => $lineItem['entity_id'],
+    ]);
+
+    if (!empty($lineItemMembership['values'][0]['membership_type_id.minimum_fee'])) {
+      $membershipMinimumFee = $lineItemMembership['values'][0]['membership_type_id.minimum_fee'];
+    }
+
+    $isOptoutUsingLastPrice = FALSE;
+    if (!empty($lineItemMembership['values'][0]["custom_$optoutUsingLastPriceFieldID"])) {
+      $isOptoutUsingLastPrice = TRUE;
+    }
+
+
+    if ($this->useMembershipLatestPrice && !$isOptoutUsingLastPrice) {
+      $unitPrice = $this->calculateSingleInstallmentAmount($membershipMinimumFee);
+    }
+    else {
+      $unitPrice = $lineItem['unit_price'];
+    }
+
+    return $unitPrice;
   }
 
+  private function calculateSingleInstallmentAmount($amount) {
+    $resultAmount =  $amount;
+    if ($this->currentInstallmentsNumber > 1) {
+      $resultAmount = floor(($amount / $this->currentInstallmentsNumber) * 100) / 100;
+    }
+
+    return $resultAmount;
+  }
 
   /**
-   * Renews the membership if
+   * Calculates the tax amount for
+   * the line item given the line item
+   * total amount and its financial type.
+   *
+   * @param float $lineTotal
+   * @param int $financialTypeId
+   *
+   * @return float
+   */
+  private function calculateLineItemTaxAmount($lineTotal, $financialTypeId) {
+    $taxAmount = 0;
+    $taxRates = CRM_Core_PseudoConstant::getTaxRates();
+    if (!empty($taxRates[$financialTypeId])) {
+      $taxRate = $taxRates[$financialTypeId];
+      $taxAmount = CRM_Contribute_BAO_Contribution_Utils::calculateTaxAmount($lineTotal, $taxRate);
+      $taxAmount = round($taxAmount['tax_amount'], 2);
+    }
+
+    return $taxAmount;
+  }
+
+  /**
+   * Sets $totalAmount and $totalTaxAmount
+   * based on the line items to be created
+   * data.
+   */
+  private function setTotalAndTaxAmount() {
+    $totalAmount = 0;
+    $taxAmount = 0;
+    foreach ($this->lineItems  as $lineItem) {
+      $totalAmount += $lineItem['line_total'] + $lineItem['tax_amount'];
+      $taxAmount += $lineItem['tax_amount'];
+    }
+
+    $this->totalAmount = $totalAmount;
+    $this->totalTaxAmount = $taxAmount;
+  }
+
+  /**
+   * Renews the payment plan and the related memberships if
    * it paid by installments.
    */
-  private function renewWithInstallmentsMembership() {
-    $previousRecurContributionId = $this->currentRecurContributionID;
+  private function renewWithInstallmentsPaymentPlan() {
     $this->createRecurringContribution();
 
+    $this->recordPaymentPlanFirstContribution();
```

Comment:
Nice approach, I hated how before we needed to instantiate `MembershipInstallmentsHandler` constructor with two different sets of input parameters to basically do the same thing. Hadn't thought of moving the logic for creating the first installment out of the class. Cool.